build: move bootloader data out of measured block

krystian-hebel opened the pull request:
Bootloader must fill some information about the kernel (address, size).
Those pieces of information are not constant. To keep hash values
independent of memory layout, those variables are moved to a separate
section out of SL (but still inside SLB). Length in SL header is no
longer set to maximum possible value, it points to the end of constant
data instead.

GRUB2 code must be changed accordingly.

Util for calculating initial PCR values after SKINIT is included
(calc_skinit_lz_sums.sh).

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp commented on Makefile:
```diff
@@ -22,8 +22,8 @@ all: lz_header.bin
 # unreferenced section ahead of .text, in which case link.lds needs adjusting.
 lz_header.bin: lz_header Makefile
 	objcopy -O binary -S --pad-to 0x10000 $< $@
```
This change probably wants splitting into a few separate patches.

First of all, introducing _sl_end means that this --pad-to can be dropped.  There should be no need to zero-extend the binary when the _sl_header changes 0xffff to _sl_end.

krystian-hebel replied:
I was afraid that GRUB would allocate what's left of SLB for something else, but even in this case, that wouldn't break anything. Pad dropped.

As a side note, it seems that the last byte of SLB cannot be measured by SKINIT even with the maximum size specified in the header.

andyhhp commented on Makefile:
```diff
@@ -22,8 +22,8 @@ all: lz_header.bin
-	@od --format=x8 --skip-bytes=4 --read-bytes=16 $@ | \
-		grep "0000004 e91192048e26f178 02ccc4765bc82a83" > /dev/null || \
+	@od --format=x8 --skip-bytes=`hexdump -s2 -n2 -e '"%u"' $@` --read-bytes=16 $@ | \
+		grep "e91192048e26f178 02ccc4765bc82a83" > /dev/null || \
```
This logic looks like it is going to be repeated often.  Seeing as cal_skinit_lz_sums is shell and can't reasonably be done in make, its probably best to add a util.sh which can be sourced by various scripts, containing a function to extract the measured length.

This od line can turn into sanity_check.sh

krystian-hebel replied:
Done

andyhhp commented on calc_skinit_lz_sums.sh:
```diff
@@ -0,0 +1,23 @@
+
+extend_sha1 () {
+	HASH=`dd if=$FILE bs=1 count=$1 2>/dev/null | sha1sum | sed "s/ .*//" \
+		| sed "s/\([a-f0-9][a-f0-9]\)/\\\\\x\1/g"`
```
These two sed's can be rather more succinctly done as:
`sed -r -e "s/ .*//g" -e "s/([a-f0-9]{2})/\\\x\1/g"`

krystian-hebel replied:
I've split this to a separate function and rewritten extend_sha* to take additional arguments. Hopefully this can be used to calculate expected PCR values at some point.

krystian-hebel committed:
setup: read size of protected mode part of kernel from zeropage

Previous code assumed existence of intermediate loader before the Linux
kernel. As it was not used, code improperly calculated hash from block
of 0 bytes.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
build: move bootloader data out of measured block

Bootloader must pass pointer to the zeropage, which is not constant. To
keep hash values independent of memory layout, those variables are moved
to a separate section out of SL (but still inside SLB). Length in SL
header is no longer set to maximum possible value, it points to the end
of constant data.

GRUB2 code must be changed accordingly.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
util: create utility functions

    extend_lz_only.sh: print expected values of PCR17 after SKINIT, after
    they are extended with SL hash only (both SHA1 and SHA256).

    extend_all.sh: takes path to bzImage and initrd (in that order) as
    arguments and calculates values expected after extending with LZ,
    kernel and initrd.

    sanity_check.sh: checks for LZ UUID at the correct offset, will be used
    by the following commit.

    util.sh: helper functions for above scripts.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

andyhhp commented on extend_all.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+. util.sh
+
+if [[ $# -ne 2 ]] || [[ ! -e "$1" ]] || [[ ! -e "$2" ]] ; then
```
`! -e` is `-z`

krystian-hebel replied:
`! -n` is `-z`, `-e` checks if it is a file. I didn't go with `-f` because I thought about using symlinks in the build scripts, now I'm not so sure about that.

andyhhp commented on extend_all.sh:
```diff
@@ -0,0 +1,14 @@
+	exit
+fi
+
+# see https://www.kernel.org/doc/html/latest/x86/boot.html#details-of-harder-fileds
```
I can't work out if the typos in this Linux's documentation are deliberate.  Normally, deliberate typos tend to be less subtle.

andyhhp commented on sanity_check.sh:
```diff
@@ -0,0 +1,7 @@
+
+od --format=x8 --skip-bytes=$SL_SIZE --read-bytes=16 $SLB_FILE | \
+	grep "e91192048e26f178 02ccc4765bc82a83" > /dev/null || \
+	{ echo "ERROR: LZ UUID missing or misplaced in $SLB_FILE" >&2; false; }
```
This, now outside of make is no longer restricted to being a oneliner.  How about:
```
if od --format=x8 --skip-bytes=$SL_SIZE --read-bytes=16 $SLB_FILE | grep -q "e91192048e26f178 02ccc4765bc82a83"; then
    echo "ERROR: LZ UUID missing or misplaced in $SLB_FILE" >&2
    false
fi
```

krystian-hebel replied:
Done

andyhhp commented on util.sh:
```diff
@@ -0,0 +1,85 @@
+		>&2 echo "\"$TRIM\" is not a valid SHA1/SHA256 hash"
+		return
+	fi
+	echo -n $TRIM | sed -r -e "s/([a-f0-9]{2})/\\\\\x\1/g"
```
You can half the number of escapes for `\x` given the `-r`

krystian-hebel replied:
Why do I still need to use 3 backslashes, is the first escape consumed already by quotation?

andyhhp commented on util.sh:
```diff
@@ -0,0 +1,85 @@
+	#echo $# >&2
+	local HASH1
+	local HASH2
+	if [ $# -eq 0 ]; then
```
`case $#`, or does that do weird integer things?

krystian-hebel replied:
Done

andyhhp commented:
No major problems.  Couple of nice-to-haves

krystian-hebel committed:
utils: post-review fixes

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
